core/services/relay/evm/chain_reader.go

```diff
@@ -0,0 +1,202 @@
+package evm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+type ChainReaderService interface {
+	services.ServiceCtx
+	commontypes.ChainReader
+}
+
+type chainReader struct {
+	lggr       logger.Logger
+	contractID common.Address
+	lp         logpoller.LogPoller
+}
+
+// NewChainReaderService constructor for ChainReader
+func NewChainReaderService(lggr logger.Logger, lp logpoller.LogPoller, ropts *types.RelayOpts) (*chainReader, error) {
+	relayConfig, err := ropts.RelayConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing RelayConfig: %w", err)
+	}
+
+	if !common.IsHexAddress(ropts.ContractID) {
+		return nil, fmt.Errorf("%w: invalid contractID, expected hex address", commontypes.ErrInvalidConfig)
+	}
+	contractID := common.HexToAddress(ropts.ContractID)
+
+	if relayConfig.ChainReader == nil {
+		return nil, fmt.Errorf("%w: ChainReader missing from RelayConfig", errors.ErrUnsupported)
+	}
+
+	if err = validateChainReaderConfig(*relayConfig.ChainReader); err != nil {
```

Comment:
This makes more sense to me than before, but I think it is still entangling things across layers. Consider if you want to test `chainReader`. You cannot just pass in your `ChainReader` config and `ContractID`. Instead, you have to construct a `RelayOpts`, which is a special type built for deserializing json encoded configs and caching the result. Why is that separate concern baked in to the constructor?

Reply:
> Instead, you have to construct a RelayOpts, which is a special type built for deserializing json

Yeah, when I wrote tests for this I did notice having to pass that in is a pain. Another option is we could just pass contractID as a separate param, I figured `RelayOpts` would be more flexible in case there's anything similar in the future that needs to get passed.  But at least for now, that's the only one and it's possible it will remain that way... we could keep our fingers crossed.